### ISSUE TITLE
02-linear-regression-medical-insurance-cost.ipynb

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/afernandez-https-github.com-4GeeksAcademy-algorithm-optimization-project-machine-learning#3]
@consultingafernandez-hub , he revisado tu proyecto. Tengo que ser directo contigo: **tienes un error CRÍTICO de data leakage** que invalida todo tu modelo. Pero también tienes cosas buenas. Vamos por partes.

---

#### 1. **Estructura Organizada** ✓
Tu notebook está bien estructurado con secciones claras (imports, carga, EDA, modelo, evaluación). Eso facilita la lectura.

#### 2. **Limpieza de Duplicados** ✓
```python
total_data = total_data.drop_duplicates().reset_index(drop=True)
```
Bien. Identificaste y eliminaste duplicados.

#### 3. **EDA Visual Completo** ✓
Histogramas, boxplots, countplots, heatmap de correlación. Buena exploración visual.

#### 4. **Guardado del Modelo** ✓
```python
dump(model, open("models/linear_regression_medical_insurance.sav", "wb"))
```
Práctica profesional: guardar el modelo para reutilizar.

---

### PROBLEMA CRÍTICO #1: DATA LEAKAGE

Mira tu código:

```python
# Paso 1: Escalas TODO el dataset
scaler = MinMaxScaler()
scal_features = scaler.fit_transform(total_data[num_variables])  # ← AQUÍ EL PROBLEMA
total_data_scal = pd.DataFrame(scal_features, ...)

# Paso 2: DESPUÉS haces el split
X_train, X_test, y_train, y_test = train_test_split(X, y, ...)
```

**¿Qué hiciste mal?**

El scaler aprende `min` y `max` de **TODO** el dataset (incluyendo lo que después será test).

```
Ejemplo simplificado:
- Dataset completo: ages = [20, 30, 40, 50, 60, 70, 80]
- Scaler aprende: min=20, max=80

- Luego splits:
  Train: [20, 30, 50, 70]
  Test:  [40, 60, 80]

- El test set "sabe" que el máximo global es 80
- Pero en producción, si llega alguien de 90 años... ¡el scaler no lo manejará bien!
```

**¿Por qué es grave?**

1. **El test set está "contaminado"** → las métricas son optimistas
2. **No simula producción** → en la realidad, no conoces datos futuros
3. **R² mentiroso** → tu modelo parece mejor de lo que realmente es

---

#### Solución:

```python
# CORRECTO: Split PRIMERO, scale DESPUÉS

# Paso 1: Split SIN escalar
X = total_data[num_variables]
y = total_data["charges"]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

#  Paso 2: Encoding (te digo en el siguiente punto)


# Paso 3: Scaler aprende SOLO de train
scaler = MinMaxScaler()
scaler.fit(X_train)  # ← Solo fit en train

# Paso 4: Transform ambos
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)  # ← Usa estadísticas de train
```

---

### Problema #2: Factorize vs One-Hot Encoding

```python
total_data["region_n"] = pd.factorize(total_data["region"])[0]
# Resultado: northeast=0, southeast=1, southwest=2, northwest=3
```

**¿Qué problema tiene esto?**

`factorize()` asigna números ordinales (0, 1, 2, 3). Esto implica que:
- northwest (3) > southwest (2) > southeast (1) > northeast (0)

**¡Pero las regiones NO tienen orden!** No es que northwest sea "más" que southeast.
**USAMOS FACTORIZE PARA VISUALIZAR LAS RELACIONES PERO NO NOS QUEDAMOS CON ELLAS SI NO QUE DESPUÉS DEL SPLIT HACEMOS EL ENCODING**

1 - Split
2 - Encoding: OPCIÓN 1: One-Hot Encoding (recomendado) / OPCIÓN 2: OneHotEncoder de sklearn (más profesional)
3 - Scaler

**Para sex y smoker** está bien usar factorize/binary porque son binarias (0/1).

---

### Problema #3: SelectKBest sin Justificación

```python
selection_model = SelectKBest(score_func=f_regression, k=4)
```

**¿Por qué k=4?** No lo explicas.

#### Qué deberías hacer:

```python
# Ver los scores de TODAS las variables
selection_model = SelectKBest(score_func=f_regression, k='all')
selection_model.fit(X_train, y_train)

# Visualizar scores
scores_df = pd.DataFrame({
    'Feature': X_train.columns,
    'Score': selection_model.scores_
}).sort_values('Score', ascending=False)

print(scores_df)

# Luego decidir k basado en los scores
# Si hay un "codo" claro en los scores, cortar ahí
```

---

### Problema #4: MUY IMPORTANTE EDA SIN CONCLUSIONES

Más que decir lo que haces en los markdowns necesito que me digas qué es lo que ves y observas en los datos y por qué tomas decisiones.
